CSHARP-2474 / CSHARP-2481: Implement libgssapi support for Kerberos auth on Linux

JamesKovacs opened the pull request:
This PR implements CSHARP-2474 and CSHARP-2481.

JamesKovacs commented:
Evergreen GSSAPI tests are passing on Windows and Linux on all supported frameworks: 
https://spruce.mongodb.com/version/6019c9e7c9ec445f87c92457/tasks

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+        private static Oid Create(params byte[] oidBytes)
+        {
+            var ntUserNameHandle = GCHandle.Alloc(oidBytes, GCHandleType.Pinned);
+            return new Oid {elements = ntUserNameHandle.AddrOfPinnedObject(), length = (uint) oidBytes.Length};
```
It's highly advisable to free the pinned handles asap. I think `Marshal.AllocHGlobal` and `Marshal.Copy` is more suitable for such unmanaged allocations.

JamesKovacs replied:
These 3 OIDs exist for the lifetime of the app domain and there can only ever be 3 OIDs due to the private ctor. I debated over whether to pin the managed byte arrays so unmanaged code could use them or whether to AllocHGlobal/Copy. Both work, but the latter will play nicer with the .NET GC. I'll switch to using AllocHGlobal/Copy. Thanks for the suggestion.

JamesKovacs replied:
Done.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:
```diff
@@ -217,7 +216,7 @@ protected sealed class SaslConversation : IDisposable
             // fields
             private readonly ConnectionId _connectionId;
-            private List<IDisposable> _itemsNeedingDisposal;
+            private ISecurityContext _securityContext;
```
minor: readonly?

JamesKovacs replied:
Not all SASL mechanisms require registering an `ISecurityContext` for disposal. Only SSPI and GSSAPI. This happens in `RegisterSecurityContext` and thus this field can't be marked `readonly`.

BorisDog replied:
Right, missed the RegisterSecurityContext.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:
```diff
@@ -247,37 +245,22 @@ public ConnectionId ConnectionId
+                }
+
+                _securityContext?.Dispose();
+                _isDisposed = true;
```
Should `_isDisposed ` functionality be threadsafe (changed by Interlocked operation), or removed at all?

JamesKovacs replied:
Dispose should only be called from a single thread and never concurrently. Even if it was called concurrently, you'd have two threads fighting to set the value of `_isDisposed` to true. Whichever won, the value would still be true. So unless I'm missing some subtlety this code should work as advertised.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/SecurityContextFactory.cs:
```diff
@@ -0,0 +1,66 @@
+/* Copyright 2020-present MongoDB Inc.
```
nit: 2021?

JamesKovacs replied:
Some of files/code were created in 2020 while others in 2021. (I have an Alfred snippets template that inserts current year in the copyright message.) I can standardize on 2021, but I don't think it matters.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/SecurityContextFactory.cs:
```diff
@@ -0,0 +1,66 @@
+                try
+                {
+                    var servicePrincipalName = $"{serviceName}/{hostname}";
+                    if (!string.IsNullOrEmpty(realm))
```
nit: IsNullOrWhitespace? Or is whitespace a valid value?

craiggwilson replied:
technically, I think whitespace is valid although I hope no one ever does that.

JamesKovacs replied:
I hope that no one ever encodes their realm name using [Whitespace](https://en.wikipedia.org/wiki/Whitespace_(programming_language)) but if they do, we have them covered.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Sspi/SspiSecurityContext.cs:
```diff
@@ -334,6 +334,17 @@ protected override bool ReleaseHandle()
 
+        protected override void Dispose(bool disposing)
+        {
+            if (!_isDisposed && disposing)
```
Should we have interlocked `_isDisposed ` pattern here as well? Or is `_credential` dispose threadsafe?

JamesKovacs replied:
We shouldn't be calling Dispose concurrently from multiple threads. Each authenticating thread will have its own SaslConversation and related ISecurityContext.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    internal static class Gss {
```
nit: Allman braces style

JamesKovacs replied:
Fixed.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+
+            uint majorStatus, minorStatus;
+            GssInputBuffer spnBuffer;
+            using (spnBuffer = new GssInputBuffer(spn))
```
minor: using (var spnBuffer = new GssInputBuffer(spn))

JamesKovacs replied:
Since `spnBuffer` must be passed by ref (see line 37), you can't declare it inside the `using` statement. If you do, you receive a C# compiler error:

```
  GssapiServicePrincipalName.cs(37, 77): [CS1657] Cannot use 'spnBuffer' as a ref or out value because it is a 'using variable'
```

JamesKovacs replied:
I've modified `GssInputBuffer` to be a class, which removed the necessity to mark them as `ref` parameters in the `gss_*` functions. This allowed me to switch to the more typical syntax of declaring the variable inside the `using`.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,112 @@
+            try
+            {
+                GssInputBuffer nameBuffer;
+                using (nameBuffer = new GssInputBuffer(username))
```
minor: `using (var nameBuffer = new GssInputBuffer(username))` or
`using var nameBuffer = new MemoryStream();` without block braces

Applies for all usings.

JamesKovacs replied:
Same as above. `using (var foo = new Foo())` doesn't play nicely with `ref` parameters.

JamesKovacs replied:
Fixed. See above.

craiggwilson commented on src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs:
```diff
@@ -340,16 +321,16 @@ public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedF
+                    bytesToSendToServer = _context.Next(bytesReceivedFromServer);
                 }
-                catch (Win32Exception ex)
+                catch (Exception ex)
```
Is there a more specific exception that can be caught here? Like, can the `LinuxException` and the `Win32Exception` have a common base class so we aren't just catching everything, but very specifically the problems that we know can go wrong in here?

JamesKovacs replied:
Fair point. I'll rework the exception handling in this portion of the code to catch either `Win32Exception` (for SASL) or `LibgssapiException` (for GSSAPI).

JamesKovacs replied:
Done.

craiggwilson commented on src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:
```diff
@@ -247,37 +245,22 @@ public ConnectionId ConnectionId
-
-            private void Dispose(bool disposing)
+            /// <param name="securityContext"></param>
+            public void RegisterSecurityContext(ISecurityContext securityContext)
```
It feels like something is leaking here. Most SASL mechanisms don't have a security context, nor do they need one, yet a conversation applies to all SASL mechanisms. Is there any reason that IDisposable isn't suitable here?

JamesKovacs replied:
This is GitHub diff weirdness. `SaslConversation` still has a `Dispose` method (see line 246). Previously there was a `RegisterItemForDisposal(IDisposable item)`, but the only `IDisposable` registered was SSPI's `SecurityContext`. So I renamed `RegisterItemForDisposal` to `RegisterSecurityContext`. Then the `Dispose` method disposes the registered `ISecurityContext` if one has been registered when the owning `SaslConversation` is disposed. The `SaslConversation` is enclosed in a `using` block (same as previously) to ensure it is properly disposed when it leaves scope.

JamesKovacs replied:
Note that `SaslConversation` called `GC.SuppressFinalize(this)` in its `Dispose` method, but it had no finalizer. So it was a NOP. Given that `SaslConversation` itself holds no unmanaged resources, there is no need for it to have a finalizer. There are other classes - such as `GssInputBuffer` - that do encapsulate unmanaged handles and I'm going to go back and plumb those with finalizers and `GC.SuppressFinalize(this)` to ensure that they're properly cleaned up if something unexpected happens.

craiggwilson replied:
Sorry, I don't think I was clear. `ISecurityContext` has no meaning whatsoever to `ScramSha1`, for instance. Except that `ScramSha1` is forced to confront `ISecurityContext` when it looks at the `SaslConversation` (think: "Am I supposed to care about this and implement it? What is it for?") The fact that `ISecurityContext` is very specific to the GSSAPI stuff means that now `ScramSha1` must purposefully choose to ignore this thing about GSSAPI.

If this is left as registering something for disposal, all `ScramSha1` needs to realize is that he has nothing that needs to be cleaned up and move on.

(This is really subtle and I might be the only one who cares about it)

JamesKovacs replied:
Fair enough point. I've switched back to the previous `RegisterItemForDisposal` impl that you wrote.

craiggwilson reviewed (changes requested):
Just a few comments/questions/suggestions. Nicely done James :)

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            if (Handle != IntPtr.Zero)
+            {
+                NativeMethods.ReleaseName(out _, Handle);
```
should it be in try/catch?

DmitryLukyanov replied:
Also, let's assign the result into `_` to explicitly say that we know that this method returns a result, but we ignore it

JamesKovacs replied:
P/invoked methods can return error codes, but can't throw exceptions. So no need to try/catch. In this case we attempt to release but there is not much we can do if the release fails. Hence ignoring the major/minor status codes.

DmitryLukyanov replied:
In expected cases, yes, but is it guaranteed? Idk this, but I can rely on you here

JamesKovacs replied:
This class was changed to derive from `SafeHandle` and a failed release of the name is communicated to .NET's diagnostic machinery.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssInputBuffer.cs:
```diff
@@ -0,0 +1,58 @@
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct GssInputBuffer : IDisposable
```
can we use this https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/NativeLibraryLoader/PinnedBuffer.cs#L21 instead?

JamesKovacs replied:
No, these are the struct layouts required by GSSAPI. Like it or lump it, these are the definitions that we have to use.

DmitryLukyanov replied:
I don't think that using structs is required. But anyway, this looks like not the solution anymore. As far as I can see, the target of these files is the same or at least very similar. So, if it's possible to reuse it, I would do it

DmitryLukyanov reviewed (commented):
Initial review, I'm going to spend a bit more time on pinvokes invistigation

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+        public IntPtr elements;
+
+        public static IntPtr NoOid = IntPtr.Zero;
+        public static Oid NtUserName = Create(0x2a, 0x86, 0x48, 0x86, 0xf7, 0x12, 0x01, 0x02, 0x01, 0x01);
```
The `Oid` struct is specifically for interop with unmanaged code. The intent is to mirror the unmanaged data structures exactly, which is why I'm using public fields and even violating .NET naming conventions (e.g. lowercase public members).

BorisDog replied:
I meant to apply this change for "helper" static fields like `NtUserName`. Which are are not part of the original structures.

JamesKovacs replied:
The OIDs must be passed by reference to the `gss_*` functions. So autoprops won't work unfortunately. I've marked the fields as read-only to prevent misuse.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssOutputBuffer.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public byte[] ToByteArray()
+        {
+            if (Length > int.MaxValue)
```
minor: should we have a lower limit than int.MaxValue? Something like 1K/1M/1G.

JamesKovacs replied:
The issue is that Length is a `uint` (based on GSSAPI struct layout) but we are allocating a managed array to hold the returned value. It shouldn't be large, but who knows what GSSAPI is going to pass back to us?

JamesKovacs replied:
Discussed via Slack. Since `length` is a 64-bit `ulong`, we have to verify that it doesn't exceed the maximum number of bytes that `Marshal.Copy` can handle.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/NativeMethods.cs:
```diff
@@ -0,0 +1,64 @@
+{
+    internal static class NativeMethods
+    {
+        private const string GSSAPI_LIBRARY = @"gssapi_krb5";
```
I debugged through this using `LD_DEBUG` as well as reading through https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html. When you use `[DllImport(LibraryName, EntryPoint="MethodName")]`, .NET uses platform dependent rules to select the library - both name and search path. On Windows, it tries `LibraryName.dll` and then `LibraryName`. So if your library name is `foo.dll`, it will first look for `foo.dll.dll` and then `foo.dll`. For Linux, the rules are `libLibraryName.so` then `LibraryName.so` then `libLibraryName` then `LibraryName`. (Not 100% sure on the ordering other than the first two because initially I used `libgssapi_krb5` and the first library it search for was `liblibgssapi_krb5.so` followed by `libgssapi_krb5.so`. Similarly on OSX where it searches for `libLibraryName.dylib`.

As long as the Linux system has `libgssapi_krb5.so` in its LD_LIBRARY_PATH, then the bindings should work as expected. And yes, we should document this adequately because it will likely require installing extra dependencies on most systems.

craiggwilson replied:
sgtm. :)

JamesKovacs commented on evergreen/evergreen.yml:
```diff
@@ -1273,9 +1314,22 @@ buildvariants:
+    - name: test-gssapi-netstandard20
+    - name: test-gssapi-netstandard21
+
+- name: gssapi-auth-tests-linux
```
I agree. The correct nomenclature is linux not posix in this case. The two places we use `posix` in `evergreen.yml` have nothing to do with my GSSAPI PR. I don't want to pollute my PR with unrelated changes.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs:
```diff
@@ -236,39 +234,24 @@ public ISaslStep Initialize(IConnection connection, SaslConversation conversatio
+                    _bytesToSendToServer = _context.Next(null);
                 }
-                catch (Win32Exception ex)
+                catch (Exception ex)
```
I reworked the exception handling in this area of the code to catch `Win32Exception` and `LibgssapiException` specifically rather than all exceptions.

```
catch (Exception ex) when (ex is Win32Exception or LibgssapiException)
```

DmitryLukyanov replied:
maybe it makes sense to add a base class here? Otherwise, we can have a 3rd exception here for mac?

JamesKovacs replied:
I assumed that `Win32Exception` was `System.ComponentModel.Win32Exception` from the .NET Framework. I didn't realize that we had defined our own. Since this is our own `Win32Exception` I agree that extracting out a common base class would be preferred to using `when`. I'll refactor the exception handling to extract a common base class. Thanks for catching this.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs:
```diff
@@ -236,39 +234,24 @@ public ISaslStep Initialize(IConnection connection, SaslConversation conversatio
+                catch (Exception ex)
                 {
-                    if (_password != null)
+                    if (password != null)
```
Actually `_password` isn't used outside of `FirstStep`. I'm going to eliminate the `_password` field and just use the passed-in `password` parameter.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/ISecurityContext.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <summary>
+        /// Negotiates the steps of an authentication handshake.
+        /// </summary>
+        /// <param name="challenge"></param>
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/ISecurityContext.cs:
```diff
@@ -0,0 +1,49 @@
+/* Copyright 2020-present MongoDB Inc.
```
This file was created in 2020. 🤷

DmitryLukyanov replied:
at someday, we probably should finally force using a `2008 - nowadays` form :)

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+
+        private static Oid Create(params byte[] oidBytes)
+        {
+            var ntUserNameHandle = GCHandle.Alloc(oidBytes, GCHandleType.Pinned);
```
I modified the OIDs to use unmanaged memory rather than pinning managed memory to avoid potential GC problems.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+        private static Oid Create(params byte[] oidBytes)
+        {
+            var ntUserNameHandle = GCHandle.Alloc(oidBytes, GCHandleType.Pinned);
+            return new Oid {elements = ntUserNameHandle.AddrOfPinnedObject(), length = (uint) oidBytes.Length};
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+*/
+
+using System;
+using System.Runtime.InteropServices;
```
Fixed.

JamesKovacs commented:
Implemented various code fixes and improvements suggested by reviewers. I believe that I've responded to everyone's comments and addressed their concerns. If I missed something, please let me know. Re-requesting reviews. Thanks in advance.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssInputBuffer.cs:
```diff
@@ -0,0 +1,71 @@
+            GC.SuppressFinalize(this);
+        }
+
+        private void InnerDispose()
```
nit: Dispose pattern usually defines two methods: `Dispose` and` Dispose(bool)`. While in this case there are no internal managed disposals, should we still follow the convention for consistency?

JamesKovacs replied:
I debated that myself. I decided that `InnerDispose` was better than `Dispose(bool disposing)` where the `disposing` parameter is never used.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssInputBuffer.cs:
```diff
@@ -0,0 +1,71 @@
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal sealed class GssInputBuffer : IDisposable
```
General thought: If `GssInputBuffer` is a short lived buffer, it would be a good candidate for pinning instead for Marshal.Alloc

JamesKovacs replied:
In most if not all cases, GssInputBuffer is a short-lived buffer for passing parameters to libgssapi. The advantage of Marshal.Alloc is that we use it for both strings and byte arrays. If we switched to pinning, we could only pin the arrays since strings have to be converted from Unicode to a C-style string. Thus we would have to remember how the buffer was created so that we could free or unpin as necessary.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+
+        protected override bool ReleaseHandle()
+        {
+            uint majorStatus = NativeMethods.ReleaseCredential(out uint minorStatus, base.handle);
```
nit: var and convert to Enums?

JamesKovacs replied:
Done. Switched to `var`. Enums won't work for the reason noted earlier.

JamesKovacs commented:
@craiggwilson: Thank you for catching that. I have corrected the other places in `GssapiAuthenticator.cs` to catch either `Win32Exception` or `LibgssapiException` rather than the more generic `Exception`.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+                    {
+                        using (var passwordBuffer = new GssInputBuffer(SecureStringHelper.ToInsecureString(password)))
+                        {
+                            majorStatus = NativeMethods.AcquireCredentialWithPassword(out minorStatus, gssName, passwordBuffer, uint.MaxValue, IntPtr.Zero, GssCredentialUsage.Initiate, out securityCredential, out OidSet _, out uint _);
```
Great catch. The returned `OidSet` is not needed. Switched to passing in `IntPtr.Zero` to avoid the allocation and need to free it. This also eliminates the need for any `OidSet` objects in the Libgssapi code.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+                    }
+                    else
+                    {
+                        majorStatus = NativeMethods.AcquireCredential(out minorStatus, gssName, uint.MaxValue, IntPtr.Zero, GssCredentialUsage.Initiate, out securityCredential, out OidSet _, out uint _);
```
Same as above. `OidSet` removed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/NativeMethods.cs:
```diff
@@ -0,0 +1,64 @@
+        public static extern uint InitializeSecurityContext(out uint minorStatus, GssapiSecurityCredential securityCredential, in IntPtr securityContextHandle, GssapiServicePrincipalName spnName, IntPtr inputMechType, GssFlags requestFlags, uint timeRequested, IntPtr inputChannelBindings, GssInputBuffer inputToken, out IntPtr actualMechType, out GssOutputBuffer outputToken, out GssFlags returnedF ...
+
+        [DllImport(GSSAPI_LIBRARY, EntryPoint = "gss_display_status")]
+        public static extern uint DisplayStatus(out uint minorStatus, uint status, GssCode statusType, in IntPtr mechType, out uint messageContext, out GssOutputBuffer statusString);
```
Unfortunately not. 0 and 1 have special meaning and hence are included in the enum, but then there are a lot of other possible failure codes. That's why `Gss.ThrowIfError()` calls into `gss_display_status` to convert arbitrary major and minor status codes into a human-readable description.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/OidSet.cs:
```diff
@@ -0,0 +1,27 @@
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct OidSet
```
Turns out we don't need it. I modified the OidSet-returning calls to request not to return OidSet objects at all and removed OidSet from our codebase.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssOutputBuffer.cs:
```diff
@@ -0,0 +1,50 @@
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct GssOutputBuffer
```
Unfortunately this cannot be a SafeHandle because libgssapi requires a very specific structure that is more than an IntPtr. As well, I could only get the libgssapi integration working when `GssOutputBuffer` is a struct. Attempting to make it a class and modifying in/out/ref calling as appropriate was not successful. So we have to be careful to call Dispose when done with a GssOutputBuffer to free its unmanaged memory.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,47 @@
+
+            if (minorStatus != 0)
+            {
+                NativeMethods.DisplayStatus(out _, minorStatus, GssCode.MECH_CODE, in Oid.NoOid, out uint _, out var outputBuffer);
```
I felt it was more intention-revealing to use `Oid.NoOid`, which is actually `IntPtr.Zero`:
```
        public static readonly IntPtr NoOid = IntPtr.Zero;
```

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssInputBuffer.cs:
```diff
@@ -0,0 +1,71 @@
+
+        private void InnerDispose()
+        {
+            if (Value != IntPtr.Zero)
```
We cannot use SafeHandle for GssInputBuffer because libgssapi requires a very specific struct layout, which we cannot provide with SafeHandle. Given that GssInputBuffer is only used to call a single libgssapi function, worst case scenario is that we don't call Dispose and the object ends up on the finalizer queue. We should never have concurrent execution of the Dispose method.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+{
+    internal class GssapiSecurityContext : SafeHandle, ISecurityContext
+    {
+        private GssapiServicePrincipalName _servicePrincipalName;
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+
+        public byte[] Next(byte[] challenge)
+        {
+            GssOutputBuffer outputToken = new GssOutputBuffer();
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// Acquires the TGT from the KDC with provided password or keytab (if password is null).
+        /// </summary>
+        /// <param name="username"></param>
```
Changed GssapiSecurityCredential to an internal class (to reduce API surface area) and removed XML comments.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <returns></returns>
+        public static GssapiSecurityCredential Acquire(string username, SecureString password)
+        {
+            IntPtr gssName = IntPtr.Zero;
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,112 @@
+        /// <summary>
+        /// Acquires the TGT from the KDC using keytab.
+        /// </summary>
+        /// <param name="username"></param>
```
Switched to internal class and removed xml comments.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/LibgssapiException.cs:
```diff
@@ -0,0 +1,37 @@
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LibgssapiException"/> class.
+        /// </summary>
+        /// <param name="majorErrorMessage"></param>
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:
```diff
@@ -247,37 +245,22 @@ public ConnectionId ConnectionId
-            }
-
-            private void Dispose(bool disposing)
+            /// <param name="securityContext"></param>
```
Switched back to previous impl using list of items registered for disposal.

JamesKovacs commented on evergreen/run-gssapi-auth-tests.sh:
```diff
@@ -0,0 +1,43 @@
+set -o errexit  # Exit the script with error if any of the commands fail
+
+# Supported/used environment variables:
+#       AUTH_HOST             Set the hostname of a key distribution center (KDC)
```
Added descriptions of environment variables used by the script that weren't documented.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssInputBuffer.cs:
```diff
@@ -0,0 +1,58 @@
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct GssInputBuffer : IDisposable
+    {
+        public ulong Length;
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssOutputBuffer.cs:
```diff
@@ -0,0 +1,50 @@
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct GssOutputBuffer
+    {
+        public ulong Length;
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+        private GssapiSecurityCredential _credential;
+        private bool _isDisposed;
+
+        public bool IsInitialized { get; private set; }
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+{
+    internal class GssapiServicePrincipalName : IDisposable
+    {
+        public IntPtr Handle { get; }
```
Changed to use SafeHandle base class instead.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+    {
+        public IntPtr Handle { get; }
+
+        public static GssapiServicePrincipalName Create(string service, string host, string realm)
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct Oid
+    {
+        public uint length;
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs:
```diff
@@ -381,14 +362,15 @@ public bool IsComplete
-                    _context.DecryptMessage(0, bytesReceivedFromServer, out decryptedBytes);
+                    // NOTE: We simply check whether we can successfully decrypt the message,
+                    //       but don't do anything with the decrypted plaintext
+                    _context.DecryptMessage(0, bytesReceivedFromServer);
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public static GssapiServicePrincipalName Create(string service, string host, string realm)
+        {
+            var spn = $"{service}@{host}";
```
Fixed.

BorisDog commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+                    }
+                    else
+                    {
+                        majorStatus = NativeMethods.AcquireCredential(out minorStatus, gssName, uint.MaxValue, IntPtr.Zero, GssCredentialUsage.Initiate, out securityCredential, IntPtr.Zero, out uint _);
```
Should `IntPtr.Zero` be passed instead of `out uint _ ` (and it all `AcquireCredential`)?

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiServicePrincipalName.cs:
```diff
@@ -0,0 +1,56 @@
+            var spn = $"{service}@{host}";
+            if (!string.IsNullOrEmpty(realm))
+            {
+                spn += $"@{realm}";
```
Verified that this required in both cases.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/SecurityContextFactory.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        public static ISecurityContext InitializeSecurityContext(ConnectionId connectionId, string serviceName, string hostname, string realm, string authorizationId, SecureString password)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,46 @@
+
+            if (majorStatus != (uint)GssStatus.Complete && majorStatus != (uint)GssStatus.ContinueNeeded)
+            {
+                NativeMethods.DisplayStatus(out _, majorStatus, GssCode.GSS_CODE, ref Oid.NoOid, out uint _, out var outputBuffer);
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssEnums.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    internal enum GssCode
+    {
+        GSS_CODE = 1,
```
These enums are serialized via their integers representations. So the names don't matter to either the C# or C code. I have renamed the values as suggested to match the libgssapi definitions.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,46 @@
+
+            if (minorStatus != 0)
+            {
+                NativeMethods.DisplayStatus(out _, minorStatus, GssCode.MECH_CODE, ref Oid.NoOid, out uint _, out var outputBuffer);
```
I'm honestly not sure why the Go driver does it that way. I haven't found a libgssapi error that required recursive lookup of the error messages. My worry is that if anything goes wrong during the error lookup, we could end up overwriting the original error with some unrelated error.

DmitryLukyanov replied:
I saw it not only on go but somewhere in docs, I can try to find it again

JamesKovacs replied:
Good catch. Certain `gss_*` functions can return a non-zero `message_context` from `gss_display_status` indicating that there is optional extra messages to follow. Modified the impl to read these extra messages if available.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,38 @@
+        public uint length;
+        public IntPtr elements;
+
+        public static IntPtr NoOid = IntPtr.Zero;
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssEnums.cs:
```diff
@@ -0,0 +1,45 @@
+
+    internal enum GssCredentialUsage
+    {
+        Both = 0,
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssEnums.cs:
```diff
@@ -0,0 +1,45 @@
+    [Flags]
+    internal enum GssFlags
+    {
+        Mutual = 2,
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,46 @@
+            {
+                NativeMethods.DisplayStatus(out _, majorStatus, GssCode.GSS_CODE, ref Oid.NoOid, out uint _, out var outputBuffer);
+                majorMessage = Marshal.PtrToStringAnsi(outputBuffer.Value);
+                outputBuffer.Dispose();
```
Refactored to ensure `GssOutputBuffer` is disposed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+            }
+            finally
+            {
+                outputBuffer.Dispose();
```
Refactored to use using.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+            }
+            finally
+            {
+                outputToken.Dispose();
```
Refactored.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityContext.cs:
```diff
@@ -0,0 +1,128 @@
+            }
+            finally
+            {
+                outputBuffer.Dispose();
```
Refactored.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    internal sealed class GssapiSecurityCredential : GssapiSafeHandle
+    {
+        public static GssapiSecurityCredential Acquire(string username, SecureString password)
```
nit: I think it would be better to split this class into two: `GssapiSecurityCredential` and `GssapiSecurityCredentialFactory`

JamesKovacs replied:
I'm mirroring the existing `SspiSecurityCredential`. If I extract out the Acquire methods into their own factory classes, then `SecurityCredentialFactory` will be calling another factory. My preference is to minimize changes to the existing SSPI impl and mirror what was done there for consistency.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs:
```diff
@@ -243,39 +244,27 @@ public ConnectionId ConnectionId
             /// <param name="item">The disposable item.</param>
             public void RegisterItemForDisposal(IDisposable item)
             {
+                Ensure.IsNotNull(item, "item");
```
nit: Ensure.IsNotNull(item, nameof(item));

JamesKovacs replied:
Good catch. I often forget that `nameof` exists for some reason. Fixed.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssOutputBuffer.cs:
```diff
@@ -0,0 +1,63 @@
+        {
+            if (value != IntPtr.Zero)
+            {
+                NativeMethods.ReleaseBuffer(out _, this);
```
nit: `_ = NativeMethods.ReleaseBuffer(out _, this);`

JamesKovacs replied:
Fixed.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/GssapiSecurityCredential.cs:
```diff
@@ -0,0 +1,75 @@
+            {
+                if (gssName != IntPtr.Zero)
+                {
+                    NativeMethods.ReleaseName(out _, gssName);
```
nit: `_ =  NativeMethods.ReleaseName(out _, gssName);`

JamesKovacs replied:
Thought I caught all these, but missed this one. Thanks. Fixed.

DmitryLukyanov reviewed (approved):
Some minor comments and then LGTM, Great job!

DmitryLukyanov commented:
NOTE: I assume that we took `DllImport` as the approach here only for historical reasons

JamesKovacs commented on evergreen/run-gssapi-auth-tests.sh:
```diff
@@ -0,0 +1,46 @@
+
+  powershell.exe .\\build.ps1 -target $TARGET
+else
+  echo "Setting krb5 config file"
```
In theory it would if P/Invoke to libgssapi worked on OSX.

DmitryLukyanov replied:
I meant that maybe it makes sense to add conditions like: 
  
    if windows
    else if linux
    else // TODO

JamesKovacs replied:
The else branch will work for Linux and OSX. I tested the technique with kinit on OSX and it worked the same as Linux.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs:
```diff
@@ -236,39 +234,22 @@ public ISaslStep Initialize(IConnection connection, SaslConversation conversatio
+            private readonly ISecurityContext _context;
 
-            public FirstStep(string serviceName, string hostName, string realm, string username, SecureString password, SaslConversation conversation)
+            public FirstStep(string serviceName, string hostname, string realm, string username, SecureString password, SaslConversation conversation)
```
Wikipedia says it is `hostname` rather than `hostName`.

BorisDog commented:
LGTM

JamesKovacs commented on evergreen/run-gssapi-auth-tests.sh:
```diff
@@ -0,0 +1,46 @@
+
+  ./build.sh -target=$TARGET
+fi;
+
```
Removed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/ISecurityContext.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <param name="challenge">Challenge received from the MongoDB server.</param>
+        /// <returns>Response to send back to the MongoDB server.</returns>
+        byte[] Next(byte[] challenge);
+        /// <summary>
```
Fixed.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Oid.cs:
```diff
@@ -0,0 +1,41 @@
+            int numBytes = oidBytes.Length;
+            var unmanagedArray = Marshal.AllocHGlobal(numBytes);
+            Marshal.Copy(oidBytes, 0, unmanagedArray, numBytes);
+            return new Oid { elements = unmanagedArray, length = (uint) numBytes };
```
Fixed.

JamesKovacs commented:
> NOTE: I assume that we took `DllImport` as the approach here only for historical reasons

I chose the `DllImport` approach for a few reasons...
1. It mirrors the existing SSPI impl.
2. I was familiar with it from previous projects.
3. While the delegate approach offers more flexibility, that flexibility wasn't required for integrating with libgssapi.

DmitryLukyanov commented:
> 
> 
> > NOTE: I assume that we took `DllImport` as the approach here only for historical reasons
> 
> I chose the `DllImport` approach for a few reasons...
> 
>     1. It mirrors the existing SSPI impl.
> 
>     2. I was familiar with it from previous projects.
> 
>     3. While the delegate approach offers more flexibility, that flexibility wasn't required for integrating with libgssapi.

got it, my point was that it would be good to be consistent in our approaches(at least latest) at least if we can and there are no disadvantages in it

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/Gss.cs:
```diff
@@ -0,0 +1,47 @@
+            string majorMessage = null;
+            string minorMessage = null;
+
+            if (majorStatus != (uint)GssStatus.GSS_S_COMPLETE && majorStatus != (uint)GssStatus.GSS_S_CONTINUE_NEEDED)
```
Done.

JamesKovacs commented on src/MongoDB.Driver.Core/Core/Authentication/Libgssapi/NativeMethods.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace MongoDB.Driver.Core.Authentication.Libgssapi
+{
+    internal static class NativeMethods
```
Switched from C# to C-style naming.

JamesKovacs commented:
Squashed and merged with `master`.
